Find in Note: Dark gray outline (shadow) appears behind gray/yellow highlights when matched text found in HTML note.

megangardner opened the pull request:
#### 89bc7dd33ce84e204457c2f54d27bafa85cf8ace
<pre>
Find in Note: Dark gray outline (shadow) appears behind gray/yellow highlights when matched text found in HTML note.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270666">https://bugs.webkit.org/show_bug.cgi?id=270666</a>
<a href="https://rdar.apple.com/122843511">rdar://122843511</a>

Reviewed by Aditya Keerthi.

In notes, the WKContentView is transparent, so our original solution of putting an additional
grey layer behind the content view that filled up the empty parts of the scroll view would show
through and make the find ui have a incorrect grey cast. So instead, we make four views that surround
the WKContentView to fill in any part of the scrollView that isn&apos;t covered by the contentView.
These are arranged around the content view like so:

----- -----------
|    |          |
|    |----------|
|    |     |    |
|    |     |    |
----- ------    |
|          |    |
|__________|____|

Each view is expanded to reach the edges of the scroll view every time the view is scrolled or the bounds change.
This means that no matter where the content view is scrolled to, there will be a view that gives the correct
grey cast to the scroll view.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidScroll:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(-[WKWebView _updateFindOverlayForOverflowScrollPositions]):
(-[WKWebView _showFindOverlay]):
(-[WKWebView _hideFindOverlay]):
(-[WKWebView _didAddLayerForFindOverlay:]):
(-[WKWebView _updateFindOverlayPosition]): Deleted.

Canonical link: <a href="https://commits.webkit.org/275873@main">https://commits.webkit.org/275873@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2c2fcab3ecd48e167271a3d7272b10b674a0d15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43126 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45526 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45752 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39249 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35648 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37150 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16646 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16786 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1180 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39323 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38537 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47295 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18040 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42441 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19577 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41100 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9596 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19755 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -201,6 +201,15 @@ struct PerWebProcessState {
 #endif // PLATFORM(IOS_FAMILY)
 
+#if HAVE(UIFINDINTERACTION)
+enum FindOverlayView {
```
Make this an enum class please?

aprotyas replied:
Oh, I see why you did it like this. My other comment speaks in a different direction if you want to consider that.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -247,7 +256,7 @@ struct PerWebProcessState {
     BOOL _findInteractionEnabled;
 #if HAVE(UIFINDINTERACTION)
-    RetainPtr<UIView> _findOverlay;
+    Vector<RetainPtr<UIView>> _findOverlaysForOverflowScroll; // Top, right, bottom, left
```
The overlay ordering enforcement is a _little_ fragile... particularly since it's documented through a comment. What do you think about introducing a structure that captures what is going on?

```cpp
struct OverflowScrollOverlays {
    RetainPtr<UIView> top;
    RetainPtr<UIView> right;
    RetainPtr<UIView> bottom;
    RetainPtr<UIView> left;
};

OverflowScrollOverlays _findOverlaysForOverflowScroll;
```

aprotyas replied:
That way you can also get rid of `enum FindOverlayView`?

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3351,40 +3351,81 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
+    auto topView = _findOverlaysForOverflowScroll[FindOverlayView::Top].get();
+    auto rightView = _findOverlaysForOverflowScroll[FindOverlayView::Right].get();
+    auto bottomView = _findOverlaysForOverflowScroll[FindOverlayView::Bottom].get();
+    auto leftView = _findOverlaysForOverflowScroll[FindOverlayView::Left].get();
```
Our smart pointer guidelines suggest these should remain as smart pointers on the stack, so I think we prefer:

```suggestion
    RetainPtr topView = _findOverlaysForOverflowScroll[FindOverlayView::Top];
    RetainPtr rightView = _findOverlaysForOverflowScroll[FindOverlayView::Right];
    RetainPtr bottomView = _findOverlaysForOverflowScroll[FindOverlayView::Bottom];
    RetainPtr leftView = _findOverlaysForOverflowScroll[FindOverlayView::Left];
```

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -201,6 +201,15 @@ struct PerWebProcessState {
 #endif // PLATFORM(IOS_FAMILY)
 
+#if HAVE(UIFINDINTERACTION)
+enum FindOverlayView {
```
should this be an enum class : uint8_t ?

rr-codes replied:
or maybe it should just be a struct instead of an enum + a loosely-ordered vector? like
```
struct FindOverlayViews {
 RetainPtr<UIView> left;
// etc
}
```

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3351,40 +3351,81 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
-        UIColor *overlayColor = [UIColor colorWithRed:(26. / 255) green:(26. / 255) blue:(26. / 255) alpha:(64. / 255)];
-        [_findOverlay setBackgroundColor:overlayColor];
+    UIColor *overlayColor = [UIColor colorWithRed:(26. / 255) green:(26. / 255) blue:(26. / 255) alpha:(64. / 255)];
+    if (!_findOverlaysForOverflowScroll.size()) {
```
Nit: `_findOverlaysForOverflowScroll.isEmpty()` is a little easier to read

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -201,6 +201,15 @@ struct PerWebProcessState {
 #endif // PLATFORM(IOS_FAMILY)
 
+#if HAVE(UIFINDINTERACTION)
+enum FindOverlayView {
```
```suggestion
enum class FindOverlayViewPosition : uint8_t {
```

Probably best to also move this to the `.mm` file, since it isn't needed outside of it.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -247,7 +256,7 @@ struct PerWebProcessState {
     BOOL _findInteractionEnabled;
 #if HAVE(UIFINDINTERACTION)
-    RetainPtr<UIView> _findOverlay;
+    Vector<RetainPtr<UIView>> _findOverlaysForOverflowScroll; // Top, right, bottom, left
```
"OverflowScroll" is bit of an overloaded term here, since this is not really about `overflow: scroll`.

`_findOverlaysOutsideContentView` might be more accurate.

aprotyas reviewed (commented):
The fix LGTM actually -- I'm suggesting an alternative approach to manage the overlays

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3351,40 +3351,81 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
-        _findOverlay = adoptNS([[UIView alloc] init]);
-        UIColor *overlayColor = [UIColor colorWithRed:(26. / 255) green:(26. / 255) blue:(26. / 255) alpha:(64. / 255)];
-        [_findOverlay setBackgroundColor:overlayColor];
+    UIColor *overlayColor = [UIColor colorWithRed:(26. / 255) green:(26. / 255) blue:(26. / 255) alpha:(64. / 255)];
```
This can be moved inside the `if`.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3351,40 +3351,81 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
     if (CALayer *contentViewFindOverlayLayer = [self _layerForFindOverlay]) {
-        [[_findOverlay layer] removeAllAnimations];
-        [contentViewFindOverlayLayer removeAllAnimations];
+        for (RetainPtr overlay : _findOverlaysForOverflowScroll) {
```
Given the repetition of this loop, we could have a method like `_updateFindOverlaysOutsideContentView:` that takes a block, and calls the block on each view.

rr-codes replied:
or just make it an initializer list like `for (RetainPtr overlay : { _findOverlaysForOverflowScroll.left, _findOverlaysForOverflowScroll.right, etc {)` ?

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3404,22 +3446,25 @@ - (void)_hideFindOverlay
     [contentViewFindOverlayLayer addAnimation:animation forKey:@"findOverlayFadeOut"];
-    [findOverlayLayer addAnimation:animation forKey:@"findOverlayFadeOut"];
+    for (RetainPtr overlay : _findOverlaysForOverflowScroll)
+        [overlay addAnimation:animation forKey:@"findOverlayFadeOut"];
```
```suggestion
        [[overlay layer] addAnimation:animation forKey:@"findOverlayFadeOut"];
```

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -247,7 +247,13 @@ struct PerWebProcessState {
     BOOL _findInteractionEnabled;
 #if HAVE(UIFINDINTERACTION)
-    RetainPtr<UIView> _findOverlay;
+    struct findOverlays {
```
nit: should be `FindOverlays`

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3404,22 +3460,31 @@ - (void)_hideFindOverlay
 
     contentViewFindOverlayLayer.opacity = 0;
-    findOverlayLayer.opacity = 0;
+    [_findOverlaysOutsideContentView.value().top layer].opacity = 0;
```
Anywhere you use `value()` should just be `->`.

But please also try to do this https://github.com/WebKit/WebKit/pull/25600#discussion_r1516993930, so that only needs to happen in one place.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3353,38 +3353,79 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
 {
     CALayer *contentViewFindOverlayLayer = [self _layerForFindOverlay];
-    CALayer *findOverlayLayer = [_findOverlay layer];
+    CALayer *findOverlayLayer = [_findOverlaysOutsideContentView.value().top layer];
```
What guarantees `_findOverlaysOutsideContentView` is not `std::nullopt` here? I think you need to update the check of line 3430 to check for that, and then move this line after the early return – otherwise this is a change in logic.

pxlcoder replied:
Also `->` rather than `value()`.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3559,20 +3608,33 @@ - (_UIDataOwner)_effectiveDataOwner:(_UIDataOwner)clientSuppliedDataOwner
 }
 
+#if HAVE(UIFINDINTERACTION)
+- (void)_updateFindOverlaysOutsideContentView:(void(^)(RetainPtr<UIView>))updateFindOverlays
```
```suggestion
- (void)_updateFindOverlaysOutsideContentView:(void(^)(UIView *))updateFindOverlays
```

pxlcoder replied:
This should be `UIView *` rather than `RetainPtr` to avoid ref-churn.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3559,20 +3608,33 @@ - (_UIDataOwner)_effectiveDataOwner:(_UIDataOwner)clientSuppliedDataOwner
+    updateFindOverlays(_findOverlaysOutsideContentView->top);
+    updateFindOverlays(_findOverlaysOutsideContentView->bottom);
+    updateFindOverlays(_findOverlaysOutsideContentView->left);
+    updateFindOverlays(_findOverlaysOutsideContentView->right);
```
```suggestion
    updateFindOverlay(_findOverlaysOutsideContentView->top);
    updateFindOverlay(_findOverlaysOutsideContentView->bottom);
    updateFindOverlay(_findOverlaysOutsideContentView->left);
    updateFindOverlay(_findOverlaysOutsideContentView->right);
```

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3353,38 +3353,79 @@ - (CABasicAnimation *)_animationForFindOverlay:(BOOL)animateIn
         [contentViewFindOverlayLayer setOpacity:1];
     } else {
-        [_findOverlay setAlpha:0];
+        if (_findOverlaysOutsideContentView) {
```
I don't think this check is necessary? The variable is always initialized by this method.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3404,22 +3445,28 @@ - (void)_hideFindOverlay
 
-        if ([strongSelf->_findOverlay alpha])
-            return;
+        [strongSelf _removeLayerForFindOverlay];
```
Shouldn't this be after the early return, as it was before?

megangardner replied:
I think this is in the right place, I added the other early return to check of the new overlays are in place before calling remove from them. It's possible, but unlikely, that we have an overlay layer without the overlay views, and I don't want to not remove the layer just because the views aren't there. (Or I could be misunderstanding your concern.)

pxlcoder replied:
It's possible to have the overlay views outside the content view without the PageOverlay layer, but not vice-versa. This is because `_showFindOverlay` always creates the overlay views outside the content view, before creating the PageOverlay.

I think the existing structure of the code is necessary in order to guard against scenarios where the find overlay is brought up again as it's animating out – which is possible due to the async nature of the code.

So the way this code needs to be written to preserve the existing structure is:

```
bool hasOverlaysOutsideContentView = strongSelf->_findOverlaysOutsideContentView.has_value();
if (hasOverlaysOutsideContentView && [strongSelf->_findOverlaysOutsideContentView->top alpha]])
    return;

[strongSelf _removeLayerForFindOverlay];
 
if (hasOverlaysOutsideContentView) {
    [strongSelf _updateFindOverlaysOutsideContentView:^(UIView *view) {
        [view removeFromSuperview];
    }];
    strongSelf->_findOverlaysOutsideContentView.reset();
}
```

megangardner committed:
Find in Note: Dark gray outline (shadow) appears behind gray/yellow highlights when matched text found in HTML note.
https://bugs.webkit.org/show_bug.cgi?id=270666
rdar://122843511

Reviewed by Aditya Keerthi.

In notes, the WKContentView is transparent, so our original solution of putting an additional
grey layer behind the content view that filled up the empty parts of the scroll view would show
through and make the find ui have a incorrect grey cast. So instead, we make four views that surround
the WKContentView to fill in any part of the scrollView that isn't covered by the contentView.
These are arranged around the content view like so:

----- -----------
|    |          |
|    |----------|
|    |     |    |
|    |     |    |
----- ------    |
|          |    |
|__________|____|

Each view is expanded to reach the edges of the scroll view every time the view is scrolled or the bounds change.
This means that no matter where the content view is scrolled to, there will be a view that gives the correct
grey cast to the scroll view.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidScroll:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(-[WKWebView _updateFindOverlayForOverflowScrollPositions]):
(-[WKWebView _showFindOverlay]):
(-[WKWebView _hideFindOverlay]):
(-[WKWebView _didAddLayerForFindOverlay:]):
(-[WKWebView _updateFindOverlayPosition]): Deleted.

Canonical link: https://commits.webkit.org/275873@main
